File: typescript/client/src/GVPlatform.ts
import axios, { AxiosInstance, AxiosResponse } from 'axios';
import { randomUUID } from 'crypto';
import { DateTime } from 'luxon';
import events from 'events';
import * as signalR from '@microsoft/signalr';
import { IPlatformNotification, NotificationEvent } from './Model';

/**
 * Class for Accessing AMPP Control Functionality
 */
export class GVPlatform extends events.EventEmitter implements NotificationEvent {
  platformUri: string;
  instance: AxiosInstance;
  apiKey: string;
  scopes: string[];
  bearerToken: string;
  notificationConnection: any;
  signal: signalR.HubConnection | any;
  correlationId: string;
  tokenRefreshTimer: any;

  /**
   * constructor
   * @param baseURL url for accessing GV Platform
   * @param apiKey The API Key (must have platform and cluster.readonly scopes)
   */
  constructor(baseURL: string, apiKey: string) {
    super();
    this.platformUri = baseURL;
    this.apiKey = apiKey;
    this.instance = axios.create({
      baseURL,
    });

    // If using AMMPP Control Only then you only need platform and cluster.readonly scopes
    // this.scopes = ['platform', 'cluster.readonly']
    // However, if you are using the Routing API then you will need the cluster scope
    this.scopes = ['platform', 'platform.readonly', 'cluster', 'cluster.readonly'];

    this.bearerToken = '';
    this.notificationConnection = null;
    this.correlationId = randomUUID();
  }

  public OnNotification(notification: IPlatformNotification) {
    this.emit('notification', notification);
  }

  /**
   * Schedule a token refresh based on its expiration time
   * The refresh will occur at 75% of expiry time
   * @param token The base64 JWT token
   */
  scheduleTokenRefresh(token: string) {
    const payloadBase64 = token.split('.')[1];
    const decodedJson = Buffer.from(payloadBase64, 'base64').toString();
    const decoded = JSON.parse(decodedJson);

    // Calculate the remaining time until expiration
    const remainingTime = decoded.exp * 1000 - Date.now();

    // Schedule a refresh at 75% of the remaining time
    const sleepTime = 0.75 * remainingTime;

    // Clear the previous timer if it exists
    clearTimeout(this.tokenRefreshTimer);

    // Set up the refresh timer.
    this.tokenRefreshTimer = setTimeout(async () => {
      await this.getToken();
    }, sleepTime);
  }

  /**
   * Obtains a JWT token from the identity service
   * Will schedule a timer to refresh the token before its expiration
   * @returns true if token obtained.
   */
  async getToken(): Promise<boolean> {
    let res: AxiosResponse;

    try {
      res = await this.instance.request({
        data: 'grant_type=client_credentials&scope=' + this.scopes.join(' '),
        headers: {
          Authorization: 'Basic ' + this.apiKey,
          'Content-Type': 'application/x-www-form-urlencoded',
        },
        method: 'POST',
        url: '/identity/connect/token',
      });

      this.bearerToken = res.data.access_token;
      this.scheduleTokenRefresh(this.bearerToken);
    } catch (err) {
      throw new Error('login error' + err);
    }

    return true;
  }

  /**
   * login
   * connects and authenticates with GVPlatform
   */
  async login(): Promise<boolean> {
    return this.getToken();
  }

  /**
   * startNotificationListener
   * Opens the Connection to the SignalR connection on the PushNotifications Service
   * And starts listening for push notifications
   * You must subscribe to topics to get events. see subscribeToNotification
   * @returns a value indicating whether the connection has been established correctly
   */
  async startNotificationListener(): Promise<boolean> {
    try {
      let hubUrl: URL = new URL('/pushnotificationshub', this.platformUri);
      let connectionBuilder = new signalR.HubConnectionBuilder()
        .withUrl(hubUrl.href, {
          accessTokenFactory: () => {
            return this.bearerToken;
          },
          transport: signalR.HttpTransportType.WebSockets,
        })
        .withAutomaticReconnect()
        .configureLogging(signalR.LogLevel.Debug)
        .withHubProtocol(new signalR.JsonHubProtocol());

      this.signal = connectionBuilder.build();

      await this.signal.start();

      // Add event handler.
      // You must subscribe to topics using subscribeToNotification() to receive notifications
      this.signal.on('ReceiveNotification', this.onNotification);

      // Test to make sure we are reciving notifications from the Hub
      this.signal.on('Pong', this.onPong);
      this.signal.invoke('Ping');
      return true;
    } catch (err) {
      console.log('error starting SignalR Connection: ', err);
    }

    return false;
  }

  /**
   * subscribeToNotification
   * @param topic the topic to listen to notifications on
   * For AMPP Control these are of the format `gv.ampp.control.{workloadId}.{command}.notify
   */
  async subscribeToNotification(topic: string) {
    const subscriptionRequest = {
      Subscriptions: [topic],
      Context: {
        CorrelationId: this.correlationId,
      },
    };
    await this.signal.invoke('Subscribe', subscriptionRequest);
  }

  async unsubscribeNotification(topic: string) {
    const subscriptionRequest = {
      Subscriptions: [topic],
      Context: {
        CorrelationId: this.correlationId,
      },
    };
    await this.signal.invoke('Unsubscribe', subscriptionRequest);
  }

  private onPong = (account: string) => {
    console.log('pong', account);
  };

  private onNotification = (notification: IPlatformNotification) => {
    this.OnNotification(notification);
  };

  public async get(url: string): Promise<AxiosResponse> {
    let res: AxiosResponse;

    res = await this.instance.request({
      headers: {
        Authorization: 'Bearer ' + this.bearerToken,
      },
      method: 'GET',
      url: url,
    });

    return res;
  }

  public async post(url: string, data: any): Promise<AxiosResponse> {
    let res: AxiosResponse;

    res = await this.instance.request({
      headers: {
        Authorization: 'Bearer ' + this.bearerToken,
      },
      method: 'POST',
      url: url,
      data,
    });

    return res;
  }

  public async put(url: string, data: any): Promise<AxiosResponse> {
    let res: AxiosResponse;

    res = await this.instance.put(url, data, {
      headers: {
        Authorization: 'Bearer ' + this.bearerToken,
        'Content-Type': 'application/json-patch+json',
        'if-match': '"*"',
      },
    });
    return res;
  }

  public async delete(url: string): Promise<AxiosResponse> {
    let res: AxiosResponse;

    res = await this.instance.delete(url, {
      headers: {
        Authorization: 'Bearer ' + this.bearerToken,
      },
    });
    return res;
  }

  /**
   * publishNotification
   * uses SignalR connection to send a notification
   * @returns a value indicating whether command has been sent
   */
  async publishNotification(topic: string, content: any): Promise<boolean> {
    var publishNotification = {
      id: randomUUID(),
      time: DateTime.utc().toISO(),
      topic,
      source: 'AMPP SDK Sample',
      ttl: 30000,
      content: JSON.stringify(content),
      contentType: null,
      contentLength: 0,
      context: {
        correlationId: this.correlationId,
      },
    };

    const response = await this.signal.invoke('PublishNotification', publishNotification);
    return response.isSuccess;
  }

  /**
   * Gets notifications for a given mailbox
   * @param mailboxId The mailbox ID to get
   * @returns A result object containing a list of notifications
   */
  public getNotifications = async (mailboxId: string) => {
    const result = await this.instance.request({
      method: 'get',
      headers: {
        'x-correlation-id': this.correlationId,
      },
      params: {
        count: 100,
        timeout: 10000,
      },
      url: `/notifications/${mailboxId}`,
    });

    return result;
  };
}

File: csharp/client/ampp.control.common/GVPlatform.cs
using AmppControl.Model;
using Microsoft.AspNetCore.SignalR.Client;
using Newtonsoft.Json.Linq;
using Newtonsoft.Json;
using Serilog;
using System;
using System.Collections.Generic;
using System.Net.Http;
using System.Text;
using System.Threading.Tasks;

namespace ampp.control.common
{
    internal class GVPlatform
    {
        private readonly HttpClient client = new HttpClient();
        private readonly string platformURL;
        private readonly string apiKey;
        private string bearerToken = null;
        private readonly string correlationId = null;
        private HubConnection hubConnection = null;


        /// <summary>
        /// Event Fired when an AMPP Control Notification Event is received
        /// </summary>
        public event EventHandler<PushNotification> OnPushNotifyEvent;

        /// <summary>
        /// Event fired when signalr hubconnection recovered from a connection loss.
        /// </summary>
        private event EventHandler<string> OnSignalRReconnected;


        public GVPlatform(string platformURL, string apiKey)
        {
            this.platformURL = platformURL;
            this.apiKey = apiKey;
            client.BaseAddress = new Uri(platformURL);
            correlationId = Guid.NewGuid().ToString();
        }

        /// <summary>
        /// Connect to GV Platform
        /// </summary>
        /// <returns></returns>
        public Task<bool> LoginAsync()
        {
            return this.GetAccessTokenAsync();
        }

        /// <summary>
        /// Starts a Connection to the SignalR Hub on the PushNotificationService to receive Notifications
        /// </summary>
        /// <returns></returns>
        public async Task<bool> StartNotificationsListenerAsync()
        {
            try
            {
                var hubConnectionUrl = $"{this.platformURL}/pushnotificationshub";

                this.hubConnection = new HubConnectionBuilder()
                    .WithUrl(hubConnectionUrl, options =>
                    {
                        options.AccessTokenProvider = () => Task.FromResult(this.bearerToken);
                    })
                    .WithAutomaticReconnect()
                    .Build();

                hubConnection.Reconnecting += HubConnection_Reconnecting;
                hubConnection.Reconnected += HubConnection_Reconnected;
                hubConnection.Closed += HubConnection_Closed;

                await hubConnection.StartAsync();

                hubConnection.On<string>("Pong", (s) => {
                    Log.Debug("Pong: " + s);
                }
                );
                await hubConnection.InvokeAsync("Ping");

                hubConnection.On<PushNotification>("ReceiveNotification", OnNotification);
                return true;
            }
            catch (Exception ex)
            {
                Log.Error($"Failed to start notification listener {ex}");
                return false;
            }
        }

        private Task HubConnection_Closed(Exception arg)
        {
            Log.Error($"SignalR hub connection closed.  Error : {arg?.Message}");
            return Task.CompletedTask;
        }

        private async Task HubConnection_Reconnected(string arg)
        {
            Log.Information($"SignalR hub reconnected.");
            OnSignalRReconnected?.Invoke(this, arg);
            await hubConnection.InvokeAsync("Ping");
        }

        /// <summary>
        /// Reconnection event handler for signalr hubconnections.
        /// </summary>
        /// <param name="arg">The cause of the disconnection</param>
        /// <returns></returns>
        private Task HubConnection_Reconnecting(Exception arg)
        {
            Log.Error($"SignalR hub connection lost. Reconnecting.  Error : {arg?.Message}");
            return Task.CompletedTask;
        }


        /// <summary>
        /// PushNotification Received from SignalR Hub
        /// </summary>
        /// <param name="notification"></param>

        private void OnNotification(PushNotification notification)
        {
            this.OnPushNotifyEvent?.Invoke(this, notification);
        }

        /// <summary>
        /// Gets an access token to be used in all REST calls using your API key
        /// Needs to be called every 30 mins to make sure the token is refreshed as it will expire after an hour
        /// </summary>
        /// <returns></returns>
        private async Task<bool> GetAccessTokenAsync()
        {
            try
            {
                var request = new HttpRequestMessage();
                request.Method = HttpMethod.Post;
                request.Headers.Add("Authorization", $"Basic {this.apiKey}");
                request.RequestUri = new Uri(this.platformURL + "/identity/connect/token");
                request.Content = new FormUrlEncodedContent(new[]
                {
                    new KeyValuePair<string, string>("grant_type", "client_credentials"),
                    // For AMPP Control you need only: platform cluster.readonly
                    // For Routing you need: platform cluster.readonly cluster
                    // new KeyValuePair<string, string>("scope", "platform cluster.readonly")
                    new KeyValuePair<string, string>("scope", "platform cluster.readonly cluster")
                });

                var result = await client.SendAsync(request);
                if (result != null && result.StatusCode == System.Net.HttpStatusCode.OK)
                {
                    // we have a valid token
                    var body = result.Content.ReadAsStringAsync().Result;
                    var data = JsonConvert.DeserializeObject<JObject>(body);
                    this.bearerToken = data.GetValue("access_token").ToString();
                    int expiryTime = data.Value<int>("expires_in");
                    ScheduleTokenRefresh(expiryTime*1000);
                    return true;
                }
                else
                {
                    Log.Error($"Failed to get access token {result?.ReasonPhrase}");
                }
            }
            catch (Exception ex)
            {
                Log.Error($"Error obtaining access token {ex.Message}");
            }
            return false;
        }

        /// <summary>
        /// Refresh the AccessToken at 75% of the remaining time until expiry or every 1 hour, whichever comes first.
        /// </summary>
        private void ScheduleTokenRefresh(int expiryTimeMs)
        {
            // Calculate 75% of the remaining time until expiry
            double remainingTime = expiryTimeMs - DateTime.Now.Millisecond;
            double interval = Math.Min(0.75 * remainingTime, TimeSpan.FromHours(1).TotalMilliseconds);

            // Create and configure the timer
            var timer = new System.Timers.Timer();
            timer.Interval = interval;
            timer.Elapsed += async (s, e) => { await GetAccessTokenAsync(); };
            timer.AutoReset = false;
            timer.Start();
        }

        /// <summary>
        /// Send Request to SignalR to Subscribe to a specific topic
        /// </summary>
        /// <param name="topic"></param>
        public Task SubscribeToNotification(string topic)
        {
            var subscriptionRequest = new
            {
                subscriptions = new string[] { topic },
                context = new { correlationId }
            };
            return this.hubConnection.InvokeAsync("Subscribe", subscriptionRequest);
        }

        public Task UnsubscribeFromNotification(string topic)
        {
            var unsubscriptionRequest = new
            {
                subscriptions = new string[] { topic },
                context = new { correlationId }
            };
            return this.hubConnection.InvokeAsync("Unsubscribe", unsubscriptionRequest);
        }

        public async Task<bool> PushNotificationAsync(string topic, JObject content)
        {
            PushNotificationRequest request = new PushNotificationRequest
            {
                Content = JsonConvert.SerializeObject(content),
                Context = new PushNotificationContext
                {
                    CorrelationId = correlationId,
                },
                Id = Guid.NewGuid().ToString(),
                Source = "AMPP SDK Sample",
                Time = DateTime.UtcNow.ToString("s"),
                Topic = topic,
                Ttl = 3000,
            };

            await hubConnection.InvokeAsync("PublishNotification", request);

            return true;
        }

        public async Task<HttpResponseMessage> Post(string url, object data)
        {
            var request = new HttpRequestMessage(HttpMethod.Post, url);
            request.Headers.Add("Authorization", $"Bearer {this.bearerToken}");


            var json = JsonConvert.SerializeObject(data);
            var content = new System.Net.Http.StringContent(json, Encoding.UTF8, "application/json");
            request.Content = content;

            var result = await client.SendAsync(request);
            return result;
        }

        public async Task<HttpResponseMessage> Get(string url)
        {
            var request = new HttpRequestMessage(HttpMethod.Get, url);
            request.Headers.Add("Authorization", $"Bearer {this.bearerToken}");
            var result = await client.SendAsync(request);
            return result;
        }

        public async Task<HttpResponseMessage> Put(string url, object data)
        {
            var request = new HttpRequestMessage(HttpMethod.Put, url);
            request.Headers.Add("Authorization", $"Bearer {this.bearerToken}");
            request.Headers.Add("if-match", "\"*\"");


            var json = JsonConvert.SerializeObject(data);
            var content = new System.Net.Http.StringContent(json, Encoding.UTF8, "application/json-patch+json");
            request.Content = content;

            var result = await client.SendAsync(request);
            return result;
        }

        public async Task<HttpResponseMessage> Delete(string url)
        {
            var request = new HttpRequestMessage(HttpMethod.Delete, url);
            request.Headers.Add("Authorization", $"Bearer {this.bearerToken}");
            var result = await client.SendAsync(request);
            return result;
        }

    }
}
